### PR TITLE
Add preClose opt + merge close and preClose hook

### LINKED ISSUE #260: Allow specifying alternative preClose / modifying close flow
### Prerequisites

- [X] I have written a descriptive issue title
- [X] I have searched existing issues to ensure the feature has not already been requested

### 🚀 Feature Proposal

The current exit flow is:

- Fastify stops receiving connections
- fastify-websocket's preClose handler is called, and all open sessions are closed
- fastify-websocket's close handler is called, and the websocket server shuts down

I'd propose to change it to something like:
- Fastify stops receiving connections
- A user-defined preClose handler is called, falling back to fastify-websocket's default preClose handler
- The preClose handler is also responsible for closing the ws server, so no need for the close handler

### Motivation

I have a usecase where I would like the websocket server to shutdown as follows:
- Stop accepting new connections
- Send a 'will shut down in X seconds' message to the active connections
- Wait either X seconds, or for all connections to be closed, then shut down the server

This is not hard to implement in my own preClose handler. I managed to get around not being able to provide my own preClose handler by registering it before registering fastify-websocketServer so it runs earlier, but that's quite hacky, and my current solution causes the close handler to error out on `server.close(done)`, because the ws server is already closed.

For context, this is the gist of my preClose hook.

```
app.addHook('preClose', async function () {
    const closeProm = new Promise(resolve => wsServer.close(resolve))
    closeProm.catch(() => {})

    for (const socket of wsServer.clients) {
      socket.send(`Server closing. Socket will shut down in ${sShutdownMargin}s`)
    }

    await Promise.race([
      new Promise(resolve => setTimeout(resolve, sShutdownMargin * 1000)),
      closeProm // If all connections close before the timeout
    ])
    
    const goingAwayCode = 1001
    for (const client of wsServer.clients) {
      client.close(goingAwayCode, 'Server is going offline')
    }

   await closeProm
```

If you're open to this feature, I'm fine with PR'ing whichever variation you deem fit. Or if you want to go a bit further, fastify-websocket could also support this kind of shutdown natively, by adding something like `sShutdownMargin` and `onShutdownMessage` options

### Example

```
fastify.register(require('@fastify/websocket'), {
  options: {
    preClose: myOwnPreClose    
  }
```

or 

```
fastify.register(require('@fastify/websocket'), {
  options: {
    sShutdownMargin: 10,
    onShutdownMessage: 'Shutting down in 10 seconds'    
  }
```

### PR DESCRIPTION
**Pull request by HDegroote**
<strike>WARNING: current master has a failing test. To get past the precommit script, I had to comment it out. Let me know if you would like a different approach.</strike>
The test in question (`Should not hijack reply for an WS request to a WS route that gets sent a normal HTTP response in a hook`) is failing on my local machine, where only 5 of the 6 planned verifications happen, but passes fine in the CI, so it does not affect this PR

All other tests (including the newly added ones) pass.

Note: I did not find a benchmark script to run as per the instructions below, so I did not check that box.

Fixes https://github.com/fastify/fastify-websocket/issues/260

#### Checklist

- [ ] run `npm run test` and `npm run benchmark`
- [X ] tests and/or benchmarks are included
- [ X] documentation is changed or added
- [ X] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

**Comment by HDegroote** [comment]
Sorry, had not realised the CI in man was broken because of something completely unrelated. The test does keep failing me as is though, both in main and in this branch. it works if I set `t.plan(5)` to `t.plan(6)`, but I don't understand the test sufficiently well to know if that makes sense

**Comment by HDegroote** [comment]
As a sanity check, could you see if the tests of current master pass on your machine?

File: README.md [hunk 1]
```diff
@@ -233,6 +233,29 @@ fastify.listen({ port: 3000 }, err => {
...
+
+By default, all ws connections are closed when the server closes. If you wish to modify this behaviour, you can pass your own `preClose` function.
+
+Note that `preClose` is responsible for closing all connections and closing the websocket server.
+
+```js
+const fastify = require('fastify')()
+
+fastify.register(require('@fastify/websocket'), {
+  preClose: (done) => {
```
Comment (mcollina) [inline README.md:247]: This should also support async functions (without the callback).

**Review by mcollina — COMMENTED** [review]
Can you amend the types? I'll check if the tests fails on main for me.

File: README.md (hunk 1 above)
Comment (HDegroote) [reply README.md]: Added a note to clarify that it indeed supports them (it supports whatever Fastify's preClose hook supports)

**Comment by HDegroote** [comment]
Added the types. Haven't worked with Typescript before, so best to sanity check. Noticed there were tests on the types, should I add something there too?

**Comment by HDegroote** [comment]
Reset that odd test to `plan(6)` again, as it was. I have a feeling this will pass the CI, because the previous run failed exactly on that test because there were too many assertions. Do note that I had to force this through without precommit check, because it fails for me. Could be something off on my end, but could also be that the test is a bit odd.

File: types/index.d.ts [hunk 2]
```diff
@@ -88,6 +89,7 @@ declare namespace fastifyWebsocket {
     errorHandler?: (this: FastifyInstance, error: Error, connection: SocketStream, request: FastifyRequest, reply: FastifyReply) => void;
     options?: WebSocketServerOptions;
     connectionOptions?: DuplexOptions;
+    preClose?: preCloseHookHandler | preCloseAsyncHookHandler;
```
Comment (mcollina) [inline types/index.d.ts:92]: can you please add a test for this typing? we use tsd.

Comment (HDegroote) [reply types/index.d.ts:92]: Done, the typings seem to be correct

**Review by mcollina — APPROVED** [review]
lgtm

**Pull request merged**